controller.go

```diff
@@ -952,29 +953,50 @@ func parseAllocateConstraintsResponse(source interface{}, machine *machine) (Con
 	if storageMatches, found := constraintsMap["storage"]; found {
 		matches := convertConstraintMatches(storageMatches)
 		for label, ids := range matches {
-			blockDevices := make([]BlockDevice, len(ids))
-			for index, id := range ids {
-				blockDevice := machine.BlockDevice(id)
-				if blockDevice == nil {
-					return empty, NewDeserializationError("constraint match storage %q: %d does not match a block device for the machine", label, id)
+			storageDevices := make([]StorageDevice, len(ids))
+			for index, storageId := range ids {
+				if id, ok := storageId.(float64); ok {
```

Comment:
Be very careful about type asserts like this as they will panic if the value is the wrong type.
We had int previously and now flot64 here. It seems like you are catering for a json marshall/unmarshal use case?
Maybe you want to use a cast float64(storageId)

Reply:
I now understand that if we have a number it's a block device and a string is a partition. Can we add a comment to explain this concept.